File: src/main/java/com/demonwav/mcdev/insight/ColorUtil.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.insight;

import com.demonwav.mcdev.platform.AbstractModuleType;
import com.demonwav.mcdev.platform.MinecraftModule;
import com.intellij.lang.ASTNode;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleUtilCore;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.JavaTokenType;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiExpression;
import com.intellij.psi.PsiExpressionList;
import com.intellij.psi.PsiIdentifier;
import com.intellij.psi.PsiLiteralExpression;
import com.intellij.psi.PsiReferenceExpression;
import com.intellij.psi.PsiType;
import java.awt.Color;
import java.util.Map;
import java.util.function.BiFunction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ColorUtil {

    @Nullable
    public static <T> T findColorFromElement(@NotNull PsiElement element,
                                             @NotNull BiFunction<Map<String, Color>, Map.Entry<String, Color>, T> function) {
        if (!(element instanceof PsiReferenceExpression)) {
            return null;
        }

        final Module module = ModuleUtilCore.findModuleForPsiElement(element);
        if (module == null) {
            return null;
        }

        MinecraftModule minecraftModule = MinecraftModule.getInstance(module);
        if (minecraftModule == null) {
            return null;
        }

        PsiReferenceExpression expression = (PsiReferenceExpression) element;
        PsiType type = expression.getType();
        if (type == null) {
            return null;
        }

        for (AbstractModuleType<?> abstractModuleType : minecraftModule.getTypes()) {
            Map<String, Color> map = abstractModuleType.getClassToColorMappings();
            for (Map.Entry<String, Color> entry : map.entrySet()) {
                // This is such a hack
                // Okay, type will be the fully-qualified class, but it will exclude the actual enum
                // the expression will be the non-fully-qualified class with the enum
                // So we combine those checks and get this
                if (entry.getKey().startsWith(type.getCanonicalText()) &&
                        entry.getKey().endsWith(expression.getCanonicalText())) {
                    return function.apply(map, entry);
                }
            }
        }
        return null;
    }

    public static void setColorTo(@NotNull PsiElement element, @NotNull String color) {
        try {
            WriteCommandAction.runWriteCommandAction(element.getProject(), () -> {
                String[] split = color.split("\\.");
                String newColorBase = split[split.length - 1];

                ASTNode node = element.getNode();
                ASTNode child = node.findChildByType(JavaTokenType.IDENTIFIER);
                if (child == null) {
                    return;
                }

                PsiIdentifier identifier = JavaPsiFacade.getElementFactory(element.getProject()).createIdentifier(newColorBase);

                child.getPsi().replace(identifier);
            });
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
    }

    public static void setColorTo(@NotNull PsiLiteralExpression expression, int value) {
        try {
            WriteCommandAction.runWriteCommandAction(expression.getProject(), () -> {
                ASTNode node = expression.getNode();

                PsiLiteralExpression literalExpression = (PsiLiteralExpression) JavaPsiFacade.getElementFactory(expression.getProject())
                        .createExpressionFromText("0x" + Integer.toHexString(value).toUpperCase(), null);

                node.getPsi().replace(literalExpression);
            });
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
    }

    public static void setColorTo(@NotNull PsiExpressionList expressionList, int red, int green, int blue) {
        try {
            WriteCommandAction.runWriteCommandAction(expressionList.getProject(), () -> {
                PsiExpression expressionOne = expressionList.getExpressions()[0];
                PsiExpression expressionTwo = expressionList.getExpressions()[1];
                PsiExpression expressionThree = expressionList.getExpressions()[2];

                ASTNode nodeOne = expressionOne.getNode();
                ASTNode nodeTwo = expressionTwo.getNode();
                ASTNode nodeThree  = expressionThree.getNode();

                PsiExpression literalExpressionOne = JavaPsiFacade.getElementFactory(expressionList.getProject())
                        .createExpressionFromText(String.valueOf(red), null);
                PsiExpression literalExpressionTwo = JavaPsiFacade.getElementFactory(expressionList.getProject())
                        .createExpressionFromText(String.valueOf(green), null);
                PsiExpression literalExpressionThree = JavaPsiFacade.getElementFactory(expressionList.getProject())
                        .createExpressionFromText(String.valueOf(blue), null);

                nodeOne.getPsi().replace(literalExpressionOne);
                nodeTwo.getPsi().replace(literalExpressionTwo);
                nodeThree.getPsi().replace(literalExpressionThree);
            });
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/demonwav/mcdev/insight/ColorUtil.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.insight

import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.util.runWriteAction
import com.intellij.openapi.module.ModuleUtilCore
import com.intellij.psi.JavaPsiFacade
import com.intellij.psi.JavaTokenType
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiExpressionList
import com.intellij.psi.PsiLiteralExpression
import com.intellij.psi.PsiReferenceExpression
import java.awt.Color

fun <T> PsiElement.findColor(function: (Map<String, Color>, Map.Entry<String, Color>) -> T): T? {
    if (this !is PsiReferenceExpression) {
        return null
    }

    val module = ModuleUtilCore.findModuleForPsiElement(this) ?: return null

    val facet = MinecraftFacet.getInstance(module) ?: return null

    val expression = this
    val type = expression.type ?: return null

    for (abstractModuleType in facet.types) {
        val map = abstractModuleType.classToColorMappings
        for (entry in map.entries) {
            // This is such a hack
            // Okay, type will be the fully-qualified class, but it will exclude the actual enum
            // the expression will be the non-fully-qualified class with the enum
            // So we combine those checks and get this
            if (entry.key.startsWith(type.canonicalText) && entry.key.endsWith(expression.canonicalText)) {
                return function(map, entry)
            }
        }
    }
    return null
}

fun PsiElement.setColor(color: String) {
    this.containingFile.runWriteAction {
        val split = color.split(".").dropLastWhile(String::isEmpty).toTypedArray()
        val newColorBase = split.last()

        val node = this.node
        val child = node.findChildByType(JavaTokenType.IDENTIFIER) ?: return@runWriteAction

        val identifier = JavaPsiFacade.getElementFactory(this.project).createIdentifier(newColorBase)

        child.psi.replace(identifier)
    }
}

fun PsiLiteralExpression.setColor(value: Int) {
    this.containingFile.runWriteAction {
        val node = this.node

        val literalExpression = JavaPsiFacade.getElementFactory(this.project)
            .createExpressionFromText("0x" + Integer.toHexString(value).toUpperCase(), null) as PsiLiteralExpression

        node.psi.replace(literalExpression)
    }
}

fun PsiExpressionList.setColor(red: Int, green: Int, blue: Int) {
    this.containingFile.runWriteAction {
        val expressionOne = this.expressions[0]
        val expressionTwo = this.expressions[1]
        val expressionThree = this.expressions[2]

        val nodeOne = expressionOne.node
        val nodeTwo = expressionTwo.node
        val nodeThree = expressionThree.node

        val facade = JavaPsiFacade.getElementFactory(this.project)

        val literalExpressionOne = facade.createExpressionFromText(red.toString(), null)
        val literalExpressionTwo = facade.createExpressionFromText(green.toString(), null)
        val literalExpressionThree = facade.createExpressionFromText(blue.toString(), null)

        nodeOne.psi.replace(literalExpressionOne)
        nodeTwo.psi.replace(literalExpressionTwo)
        nodeThree.psi.replace(literalExpressionThree)
    }
}
